src/emu_os_bip32.c

```diff
@@ -121,128 +205,266 @@ static size_t get_seed_from_env(const char *name, uint8_t *seed, size_t max_size
   return seed_size;
 }
 
-unsigned long sys_os_perso_derive_node_bip32(cx_curve_t curve, const uint32_t *path, size_t length, uint8_t *private_key, uint8_t* chain)
+static int hdw_bip32_ed25519(extended_private_key *key, const uint32_t *path, size_t length, uint8_t *private_key, uint8_t* chain)
 {
-  uint8_t tmpPrivateChain[CX_SHA512_SIZE + 32];
-  cx_ecfp_256_public_key_t tmpEcPublic;
-  cx_ecfp_private_key_t tmpEcPrivate;
   const cx_curve_domain_t *domain;
-  uint8_t seed[64], tmp[1+64+4];
-  ssize_t seed_size;
+  cx_ecfp_256_public_key_t pub;
+  uint8_t *kP, *kLP, *kRP;
+  unsigned int i, j, len;
+  uint8_t tmp[1+64+4], x;
+  uint8_t *ZR, *Z, *ZL;
+  uint8_t a[33], b[33];
+
+  Z = tmp + 1;
+  ZL = tmp + 1;
+  ZR = tmp + 1 + 32;
+
+  kP = key->private_key;
+  kLP = key->private_key;
+  kRP = &key->private_key[32];
+
+  domain = cx_ecfp_get_domain(CX_CURVE_Ed25519);
+  pub.W_len = 65;
+  pub.W[0] = 0x04;
+
+  for (i = 0; i < length; i++) {
+    // compute the public key A = kL.G
+    memcpy(pub.W + 1, domain->Gx, 32);
+    memcpy(pub.W + 1 + 32, domain->Gy, 32);
+
+    le2be(kLP, 32);
+    sys_cx_ecfp_scalar_mult(CX_CURVE_Ed25519, pub.W, pub.W_len, kLP, 32);
+    be2le(kLP, 32);
+
+    sys_cx_edward_compress_point(CX_CURVE_Ed25519, pub.W, pub.W_len);
+
+    //Step 1: compute kL/Kr child
+    //  if less than 0x80000000 => setup 02|A|i in tmp
+    if (((path[i] >> 24) & 0x80) == 0) {
+      tmp[0] = 0x02;
+      memcpy(tmp + 1, &pub.W[1], 32);
+      len = 1 + 32;
+    } else {
+      //  else if greater-eq than 0x80000000 --> setup 00|k|i in tmp
+      tmp[0] = 0x00;
+      memcpy(tmp + 1, kP, 64);
+      len = 1 + 32 * 2;
+    }
+
+    tmp[len+0] = (path[i] >> 0) & 0xff;
+    tmp[len+1] = (path[i] >> 8) & 0xff;
+    tmp[len+2] = (path[i] >> 16) & 0xff;
+    tmp[len+3] = (path[i] >> 24) & 0xff;
+    len += 4;
+
+    // compute Z = Hmac(...)
+    cx_hmac_sha512(key->chain_code, 32, tmp, len, Z, CX_SHA512_SIZE);
+    // kL = 8*Zl + kLP (use multm, but never overflow order, so ok)
+    le2be(ZL, 32);
+    le2be(kLP, 32);
+    memset(ZL, 0, 4);
+    cx_math_add(ZL, ZL, ZL, 32);
+    cx_math_add(ZL, ZL, ZL, 32);
+    cx_math_add(ZL, ZL, ZL, 32);
+    cx_math_add(ZL, ZL, kLP, 32);
+    be2le(ZL, 32);
+    be2le(kLP, 32);
+    // kR = Zr + kRP
+    le2be(ZR, 32);
+    le2be(kRP, 32);
+
+    // cx_math_add(ZR, ZR, kRP, 32) can return invalid result when an overflow
+    // occurs
```

Comment:
The comment would be clearer if something like this is added: " so perform addition on 33 bytes and truncate the result to 32 bytes"

Reply:
Fixed. Thanks!